### PR TITLE
Add list_locks

### LINKED ISSUE #395: Add list_locks task to CLI
The CLI currently has the ability to list the current tasks, jobs, and queues, and allows to filter each of these by the relevant locks, but does not provide a way to list the current (or historical) locks.  Adding the ability to list these would simplify daily admin of projects using Procrastinate

### PR DESCRIPTION
**Pull request by ewjoachim**
Closes #395 
Because of #400 , it's not the cleanest PR ever, but the PR that will handle #400 will fix this.

Also, an ORM would definitely help.

### Successful PR Checklist:

- [x] Tests
  - [ ] (not applicable?)
- [x] Documentation
  - [ ] (not applicable?)

### CODE REVIEW

**Comment by ewjoachim** [comment]
👋 @elemoine

File: procrastinate/manager.py [hunk 1]
```diff
@@ -423,9 +423,60 @@ async def list_tasks_async(
...
 
+    async def list_locks_async(
+        self,
+        queue: Optional[str] = None,
+        task: Optional[str] = None,
+        status: Optional[str] = None,
+        lock: Optional[str] = None,
+    ) -> Iterable[Dict[str, Any]]:
+        """
+        List all locks and number of jobs per status for each lock.
```
Comment (elemoine) [inline procrastinate/manager.py:434]: Do you mean "for each task" here?

File: procrastinate/sql/queries.sql [hunk 2]
```diff
@@ -149,3 +149,35 @@ SELECT task_name AS name,
...
+          lock,
+          args,
+          status,
+          scheduled_at,
+          attempts
+     FROM procrastinate_jobs
+    WHERE (%(queue_name)s IS NULL OR queue_name = %(queue_name)s)
+      AND (%(task_name)s IS NULL OR task_name = %(task_name)s)
+      AND (%(status)s IS NULL OR status = %(status)s)
+      AND (%(lock)s IS NULL OR lock = %(lock)s)
```
Comment (elemoine) [inline procrastinate/sql/queries.sql:168]: When https://github.com/peopledoc/procrastinate/pull/402 is merged I guess we don't want to have NULL locks in the list returned by list_locks, do we?

File: procrastinate/sql/queries.sql [hunk 3]
```diff
@@ -149,3 +149,35 @@ SELECT task_name AS name,
   FROM jobs AS j
  GROUP BY name
  ORDER BY name;
+
+-- list_locks --
+-- Get list of locks and number of jobs per task
```
Comment (ewjoachim) [inline procrastinate/sql/queries.sql:154]:
```suggestion
-- Get list of locks and number of jobs per lock
```

File: procrastinate/sql/queries.sql (hunk 2 above)
Comment (ewjoachim) [reply procrastinate/sql/queries.sql:168]: I believe it's already the case ? If `%(lock)s IS NULL`, we don't filter on lock.

Comment (ewjoachim) [reply procrastinate/sql/queries.sql:168]: Ah you mean
```suggestion
      AND (%(lock)s IS NULL OR lock = %(lock)s) AND lock IS NOT NULL
```

File: procrastinate/manager.py [hunk 4]
```diff
@@ -423,9 +423,60 @@ async def list_tasks_async(
...
+        for row in await self.connector.execute_query_all_async(
+            query=sql.queries["list_locks"],
+            queue_name=queue,
+            task_name=task,
+            status=status,
+            lock=lock,
+        ):
+            if row["name"] is None:
+                # https://github.com/peopledoc/procrastinate/issues/400
+                continue
```
Comment (ewjoachim) [inline procrastinate/manager.py:463]: These 3 lines should be remove when 402 is merged

File: procrastinate/sql/queries.sql (hunk 2 above)
Comment (elemoine) [reply procrastinate/sql/queries.sql:168]: Yes, exactly.

File: procrastinate/sql/queries.sql [hunk 5]
```diff
@@ -149,3 +149,35 @@ SELECT task_name AS name,
...
+          status,
+          scheduled_at,
+          attempts
+     FROM procrastinate_jobs
+    WHERE (%(queue_name)s IS NULL OR queue_name = %(queue_name)s)
+      AND (%(task_name)s IS NULL OR task_name = %(task_name)s)
+      AND (%(status)s IS NULL OR status = %(status)s)
+      AND (%(lock)s IS NULL OR lock = %(lock)s)
+)
+SELECT lock AS name,
```
Comment (elemoine) [inline procrastinate/sql/queries.sql:170]: Super minor but I'd either remove the alias (`AS name`) or use `lock_name` as the alias (`AS lock_name`).

File: procrastinate/sql/queries.sql [hunk 6]
```diff
@@ -149,3 +149,35 @@ SELECT task_name AS name,
...
+                  COUNT(*) AS jobs_count
+             FROM jobs
+            WHERE lock = j.lock
+            GROUP BY status
+           )
+           SELECT json_object_agg(status, jobs_count) FROM stats
+       ) AS stats
+  FROM jobs AS j
+ GROUP BY name
+ ORDER BY name;
```
Comment (elemoine) [inline procrastinate/sql/queries.sql:183]: Don't you think that the following query is simpler?

```sql
WITH jobs AS (
  SELECT
    id,
    lock,
    status
  FROM procrastinate_jobs
  WHERE (%(queue_name)s IS NULL OR queue_name = %(queue_name)s)
  AND (%(task_name)s IS NULL OR task_name = %(task_name)s)
  AND (%(status)s IS NULL OR status = %(status)s)
  AND (%(lock)s IS NULL OR lock = %(lock)s)
  AND lock IS NOT NULL
), locks AS (
  SELECT
    lock,
    status,
    count(id) AS jobs_count
  FROM jobs
  GROUP BY lock, status
)
SELECT
  lock AS name,
  sum(jobs_count) AS jobs_count,
  json_object_agg(status, jobs_count)
FROM locks
GROUP BY name
ORDER BY name;
```

File: procrastinate/sql/queries.sql [hunk 7]
```diff
@@ -149,3 +149,35 @@ SELECT task_name AS name,
...
+-- Get list of locks and number of jobs per lock
+WITH jobs AS (
+   SELECT id,
+          queue_name,
+          task_name,
+          lock,
+          args,
+          status,
+          scheduled_at,
+          attempts
```
Comment (elemoine) [inline procrastinate/sql/queries.sql:163]: You just need to select the `id`, `lock` and `status` columns here.

**Comment by elemoine** [comment]
This looks great @ewjoachim. I just added a few comments that we may want to take into account.

File: procrastinate/manager.py (hunk 1 above)
Comment (ewjoachim) [reply procrastinate/manager.py:434]: Nope, it's grouped by lock value, no?

**Pull request merged**